packages/rolldown/src/utils/compose-js-plugins.ts

```diff
@@ -0,0 +1,276 @@
+import { Plugin, RolldownPlugin } from '../plugin'
+import _ from 'lodash-es'
+import { normalizeHook } from './normalize-hook'
+import { isNullish } from './misc'
+import { BuiltinPlugin } from '../plugin/builtin-plugin'
+
+const unsupportedHooks = new Set([
+  'augmentChunkHash',
+  'banner',
+  'generateBundle',
+  'moduleParsed',
+  'onLog',
+  'options',
+  'outputOptions',
+  'renderError',
+  'renderStart',
+  'resolveDynamicImport',
+  'writeBundle',
+])
+
+function createComposedPlugin(plugins: Plugin[]): Plugin {
+  // Throw errors if we try to merge plugins with unsupported hooks
+
+  const names: string[] = []
+  const batchedHooks: {
+    [K in keyof Plugin]?: NonNullable<Plugin[K]>[]
+  } = {}
+
+  plugins.forEach((plugin, index) => {
+    const pluginName = plugin.name || `Anonymous(index: ${index})`
+    names.push(pluginName)
+    ;(Object.keys(plugin) as (keyof typeof plugin)[]).forEach((pluginProp) => {
+      switch (pluginProp) {
+        case 'name':
+          break
+        case 'buildStart': {
+          const handlers = batchedHooks.buildStart ?? []
+          batchedHooks.buildStart = handlers
+          if (plugin.buildStart) {
+            handlers.push(plugin.buildStart)
+          }
+          break
+        }
+        case 'load': {
+          const handlers = batchedHooks.load ?? []
+          batchedHooks.load = handlers
+          if (plugin.load) {
+            handlers.push(plugin.load)
+          }
+          break
+        }
+        case 'transform': {
+          const handlers = batchedHooks.transform ?? []
+          batchedHooks.transform = handlers
+          if (plugin.transform) {
+            handlers.push(plugin.transform)
+          }
+          break
+        }
+        case 'resolveId': {
+          const handlers = batchedHooks.resolveId ?? []
+          batchedHooks.resolveId = handlers
+          if (plugin.resolveId) {
+            handlers.push(plugin.resolveId)
+          }
+          break
+        }
+        case 'buildEnd': {
+          const handlers = batchedHooks.buildEnd ?? []
+          batchedHooks.buildEnd = handlers
+          if (plugin.buildEnd) {
+            handlers.push(plugin.buildEnd)
+          }
+          break
+        }
+        case 'renderChunk': {
+          const handlers = batchedHooks.renderChunk ?? []
+          batchedHooks.renderChunk = handlers
+          if (plugin.renderChunk) {
+            handlers.push(plugin.renderChunk)
+          }
+          break
+        }
+        case 'augmentChunkHash':
+        case 'banner':
+        case 'generateBundle':
+        case 'moduleParsed':
+        case 'onLog':
+        case 'options':
+        case 'outputOptions':
+        case 'renderError':
+        case 'renderStart':
+        case 'resolveDynamicImport':
+        case 'writeBundle': {
+          throw new Error(
```

Comment:
- Do we plan to support more of these in the future, or are some of them not possible to compose?
- Is it possible to compose as much as we can and only skip the incompatible plugins?

Reply:
> Do we plan to support more of these in the future, or are some of them not possible to compose?

Yeah. I don't see any hook which is not composable yet. There are going to be supported in the future.

> Is it possible to compose as much as we can and only skip the incompatible plugins?

Yeah. We are doing it now. https://github.com/rolldown/rolldown/blob/8aad7d0fc33c4ca991352b86362b0c0fa383944c/packages/rolldown/src/utils/compose-js-plugins.ts#L250-L275